azure/services/vmextensions/vmextensions.go

```diff
@@ -55,6 +55,11 @@ func (s *Service) Reconcile(ctx context.Context) error {
 	defer span.End()
 
 	for _, extensionSpec := range s.Scope.VMExtensionSpecs() {
+		if _, err := s.client.Get(ctx, s.Scope.ResourceGroup(), extensionSpec.VMName, extensionSpec.Name); err == nil {
```

Comment:
This is an interesting choice. The PR makes sense, I wonder why we wouldn't want to check/apply the extension with idempotency inside the VM reconciler? Maybe I missed some discussion on that, or wdyt?

Reply:
it came out of [this conversation](https://kubernetes.slack.com/archives/CEX9HENG7/p1614873280120900) on slack. The extension was getting applied on every reconcile loop which, for VMs isn't a big deal, but for VMSS was causing the instances to always be in latest model "false" status since the extension was applied on the scale set after the VMSS VMs were created.

This also gets us closer to the behavior that will be implemented soon for #603 (PR incoming) since we'll have to do a GET to fetch the extension status and set the corresponding condition. Separated this fix into this PR so we can backport it to alpha3.

Reply:
but why reconcile the VMSS extension separately from the VMSS at all? In this PR I see you moved extension generation into the scaleset service and no-op'd the check here. Can we simply remove this code? https://github.com/CecileRobertMichon/cluster-api-provider-azure/blob/65b62e65c90a590fb362919d24317838bb05603e/exp/controllers/azuremachinepool_reconciler.go#L72-L74

And remove the vmss extensions service altogether? I think that's pretty much what AKS does, fwiw, and I don't think we hit the same issue with the extension being applied repeatedly. To get the extension status I _think_ you'll need the per instance view anyway.
 
relatively minor, not really worth blocking on, just thought I'd mention it.

Reply:
hmm, I see the extensions client actually does return the status. I was assuming we'd pluck it from the VM/VMSS instance view, which appears to have the same data:

https://pkg.go.dev/github.com/Azure/azure-sdk-for-go@v52.3.1+incompatible/services/compute/mgmt/2020-12-01/compute#VirtualMachineInstanceView

Reply:
> And remove the vmss extensions service altogether? I think that's pretty much what AKS does, fwiw, and I don't think we hit the same issue with the extension being applied repeatedly. To get the extension status I think you'll need the per instance view anyway. I think you can do the same for individual VM too.

I see what you're saying now. I think you're right. I want to keep this PR with as little LOC changes as possible to facilitate backport but I'll look into this as part of #603.